File: domain/src/test/kotlin/com/kyovo/domain/service/AuthServiceTest.kt
package com.kyovo.domain.service

import com.kyovo.domain.exception.EmailAlreadyUsedException
import com.kyovo.domain.exception.InvalidCredentialsException
import com.kyovo.domain.model.*
import com.kyovo.domain.port.secondary.PasswordHashPort
import com.kyovo.domain.port.secondary.UserRepository
import com.kyovo.domain.provider.TimeProvider
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.mockito.kotlin.*
import java.time.LocalDateTime
import java.util.*

class AuthServiceTest
{
    private val userRepository: UserRepository = mock()
    private val passwordHashPort: PasswordHashPort = mock()
    private val timeProvider = object : TimeProvider
    {
        override fun now(): LocalDateTime
        {
            return LocalDateTime.of(2026, 4, 1, 0, 0)
        }
    }
    private val authService = AuthService(userRepository, passwordHashPort, timeProvider)

    private val userId = UserId(UUID.randomUUID())
    private val existingUser = User(
        userId,
        UserName("Alice"),
        UserEmail("alice@example.com"),
        UserPassword("hashed"),
        UserRole.USER,
        UserRegistrationDate(timeProvider.now())
    )

    @Test
    fun `register hashes the password before persisting`()
    {
        val newUser = NewUser(UserName("Alice"), UserEmail("alice@example.com"), UserPassword("plain"))
        val captor = argumentCaptor<User>()
        whenever(userRepository.findByEmail(newUser.email)).thenReturn(null)
        whenever(passwordHashPort.hash("plain")).thenReturn(UserPassword("hashed"))
        whenever(userRepository.save(any())).thenAnswer { it.getArgument<User>(0) }

        authService.register(newUser)

        verify(userRepository).save(captor.capture())
        assertThat(captor.firstValue.password).isEqualTo(UserPassword("hashed"))
    }

    @Test
    fun `register returns the user persisted by the repository`()
    {
        val newUser = NewUser(UserName("Alice"), UserEmail("alice@example.com"), UserPassword("plain"))
        whenever(userRepository.findByEmail(newUser.email)).thenReturn(null)
        whenever(passwordHashPort.hash("plain")).thenReturn(UserPassword("hashed"))
        whenever(userRepository.save(any())).thenReturn(existingUser)

        val result = authService.register(newUser)

        assertThat(result).isEqualTo(existingUser)
    }

    @Test
    fun `register throws EmailAlreadyUsedException when email is taken`()
    {
        val newUser = NewUser(UserName("Dup"), UserEmail("alice@example.com"), UserPassword("plain"))
        whenever(userRepository.findByEmail(newUser.email)).thenReturn(existingUser)

        assertThatThrownBy { authService.register(newUser) }
            .isInstanceOf(EmailAlreadyUsedException::class.java)
    }

    @Test
    fun `login returns the user when credentials are valid`()
    {
        whenever(userRepository.findByEmail(UserEmail("alice@example.com"))).thenReturn(existingUser)
        whenever(passwordHashPort.matches("plain", UserPassword("hashed"))).thenReturn(true)

        val result = authService.login(UserEmail("alice@example.com"), "plain")

        assertThat(result).isEqualTo(existingUser)
    }

    @Test
    fun `login throws InvalidCredentialsException when email is unknown`()
    {
        whenever(userRepository.findByEmail(any())).thenReturn(null)

        assertThatThrownBy { authService.login(UserEmail("unknown@example.com"), "plain") }
            .isInstanceOf(InvalidCredentialsException::class.java)
    }

    @Test
    fun `login throws InvalidCredentialsException when password is wrong`()
    {
        whenever(userRepository.findByEmail(UserEmail("alice@example.com"))).thenReturn(existingUser)
        whenever(passwordHashPort.matches("wrong", UserPassword("hashed"))).thenReturn(false)

        assertThatThrownBy { authService.login(UserEmail("alice@example.com"), "wrong") }
            .isInstanceOf(InvalidCredentialsException::class.java)
    }
}

File: frontend/src/domain/service/AuthService.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest'
import { AuthService } from '@/domain/service/AuthService'
import type { AuthRepository } from '@/domain/port/secondary/AuthRepository'
import type { User } from '@/domain/model/User'

const mockUser: User = {
  id: 'user-1',
  name: 'Alice',
  email: 'alice@example.com',
  role: 'USER',
  registeredAt: '2024-01-01T00:00:00Z',
  statusInfo: { status: 'ACTIVE', since: '2024-01-01T00:00:00Z', reason: null },
}

describe('AuthService', () => {
  let repository: AuthRepository
  let service: AuthService

  beforeEach(() => {
    repository = {
      login: vi.fn().mockResolvedValue({ user: mockUser, token: 'mock-token' }),
      register: vi.fn().mockResolvedValue(mockUser),
    }
    service = new AuthService(repository)
  })

  it('delegates login to the repository and returns user and token', async () => {
    const result = await service.login({ email: 'alice@example.com', password: 'secret' })
    expect(repository.login).toHaveBeenCalledWith({ email: 'alice@example.com', password: 'secret' })
    expect(result.user).toEqual(mockUser)
    expect(result.token).toBe('mock-token')
  })

  it('delegates register to the repository and returns the created user', async () => {
    const user = await service.register({ name: 'Alice', email: 'alice@example.com', password: 'secret' })
    expect(repository.register).toHaveBeenCalledOnce()
    expect(user).toEqual(mockUser)
  })
})
